File: solutions/ruby/problem-00026.rb
#!/usr/bin/env ruby
# [minhuang@mocha:ruby]$ time ruby problem-00026.rb 
# 983
#
# real  0m0.032s
# user  0m0.016s
# sys 0m0.008s
class Array
  def tail
    self.drop(1)
  end
end

def order(a, m)
  b = 1
  (1...m).each do |i|
    b = (a*b)%m
    return i if b == 1
  end
end

def max_period(best, numbers)
  return best if numbers.nil?
  n = numbers.first
  ord = order(10, n)
  return [n, ord] if ord == n - 1
  return max_period([n, ord], numbers.tail) if ord > best.last
  return max_period(best, numbers.tail)
end

numbers = 999.step(2, -1).select { |n| n % 2 != 0 && n % 5 != 0 }
result = max_period([0, 0], numbers)
puts result.first

File: src/scala/problem-00026.scala
/*
[minhuang@mocha:scala]$ time scala problem-00026.scala 
983

real  0m1.105s
user  0m0.252s
sys 0m0.048s
*/
import scala.Math._

implicit def tupleWrapper(t: Pair[Int, Int]) = new {
  def n = t._1
  def ord = t._2
}

def order(a: Int, m: Int) = {
  def recur(b: Int, numbers: List[Int]): Int = numbers match {
    case x :: Nil => x
    case x :: xs if b == 1 => x
    case _ => recur((a * b) % m, numbers.tail)
  }
  recur(a % m, (1 to m - 1).toList)
}

val numbers = (999 to 2 by -1).filter(n => n % 2 != 0 && n % 5 != 0).toList
def maxPeriod(best: Pair[Int, Int], numbers: List[Int]): Pair[Int, Int] =
    numbers match {
  case Nil => best
  case _ => {
    val n = numbers.first
    order(10, n) match {
      case ord if ord == n - 1 => (n, ord)
      case ord if ord > best.ord => maxPeriod((n, ord), numbers.tail)
      case _ => maxPeriod(best, numbers.tail)
    }
  }
}

val result = maxPeriod((0, 0), numbers)
println(result.n)
